speechbrain/k2_integration/lexicon.py

```diff
@@ -0,0 +1,538 @@
+"""Lexicon class and utilities. Provides functions to read/write
+lexicon files and convert them to k2 ragged tensors. The Lexicon
+class provides a way to convert a list of words to a ragged tensor
+containing token IDs. It also stores the lexicon graph which can
+be used by a graph compiler to decode sequences.
+
+This code was adjusted, and therefore heavily inspired or taken from
+from icefall's (https://github.com/k2-fsa/icefall) Lexicon class and
+its utility functions.
+
+
+Authors:
+  * Zeyu Zhao 2023
+  * Georgios Karakasidis 2023
+"""
+
+
+import logging
+import re
+import sys
+import os
+from pathlib import Path
+from typing import List, Tuple, Union
+
+try:
+    import k2
+except ImportError:
+    MSG = "Please install k2 to use k2 training \n"
+    MSG += "E.G. run: pip install k2\n"
+    raise ImportError(MSG)
+
+import torch
+
+logger = logging.getLogger(__name__)
+
+
+def get_lexicon(lang_dir, csv_files, extra_vocab_files, add_word_boundary=True):
+    """Read csv_files to generate a $lang_dir/lexicon.txt for k2 training.
+    This usually includes the csv files of the training set and the dev set in the
+    output_folder. During training, we need to make sure that the lexicon.txt contains
+    all (or the majority of) the words in the training set and the dev set.
+
+    NOTE: This assumes that the csv files contain the transcription in the last column.
+
+    Also note that in each csv_file, the first line is the header, and the remaining
+    lines are in the following format:
+
+    ID, duration, wav, spk_id, wrd (transcription)
+
+    We only need the transcription in this function.
+
+    Writes out $lang_dir/lexicon.txt
+
+    Note that the lexicon.txt is a text file with the following format:
+    word1 phone1 phone2 phone3 ...
+    word2 phone1 phone2 phone3 ...
+
+    In this code, we simply use the characters in the word as the phones.
+    You can use other phone sets, e.g., phonemes, BPEs, to train a better model.
+
+    Arguments
+    ---------
+    lang_dir: str
+        The directory to store the lexicon.txt
+    csv_files: List[str]
+        A list of csv file paths
+    extra_vocab_files: List[str]
+        A list of extra vocab files. For example, for librispeech this could be the
+        librispeech-vocab.txt file.
+    add_word_boundary: bool
+        whether to add word boundary symbols <eow> at the end of each line to the
+        lexicon for every word.
+
+    Example
+    -------
+    >>> from speechbrain.k2_integration.lexicon import get_lexicon
```

Comment:
This is a note for @Adel-Moumen and @mravanelli, but could be of interest for @ZhaoZeyu1995 as well. Great that we have examples now for this, but we also have to exclude them from the testing suit because we must install k2 to execute them. Do you think we could have a way of installing k2 in the CI to be able to also run these tests? the k2_integration folder contains quite a lot of classes, we might want to test them ..

Reply:
I also see that we now have a test_k2.py file. Meaning that we really somehow want to have k2 installed in the CI. Adel any idea?

Reply:
Sorry, I am not sure how to install k2 in the CI but it will be definitely great for the tests if we can really do that.

Reply:
> Sorry, I am not sure how to install k2 in the CI but it will be definitely great for the tests if we can really do that.

I suppose you want to install a CPU version of k2 in the CI since github actions do not provide GPUs.

The steps to install k2 inside the CI are super simple. 

Take Linux as an example,

```
# First, install torch
pip install torch==2.0.1+cpu -f https://download.pytorch.org/whl/torch_stable.html

# Second, install k2
pip install k2==1.24.3.dev20230719+cpu.torch2.0.1 -f https://k2-fsa.github.io/k2/cpu.html

# Done!
```

Please refer to https://k2-fsa.github.io/k2/installation/from_wheels.html#linux-cpu-example
for details.

By the way, you have to first choose which version of PyTorch to install. After that, you can install the corresponding version of k2 that is compiled against the installed version of PyTorch.

Reply:
You can find the list of CPU wheels for k2 at
https://k2-fsa.github.io/k2/cpu.html

Please always use the latest version of k2, which is backward compatible but contains bug fixes.

Reply:
If you want to use a `requirements.txt` to install k2, please refer to
https://github.com/k2-fsa/icefall/blob/master/requirements-ci.txt

```
-f https://download.pytorch.org/whl/cpu/torch_stable.html torch==1.13.1+cpu
-f https://k2-fsa.org/nightly/ k2==1.23.4.dev20230319+cpu.torch1.13.1
```
You only need the above two lines in `requirements.txt` to install a CPU version of k2.

Please switch to the latest version of k2 if possible. The above example does not use the latest version of k2.
Also, please replace
```
-f https://k2-fsa.github.io/k2/cpu.html
```
with
```
-f https://k2-fsa.github.io/k2/cpu.html
```

Reply:
Hello @csukuangfj, 

Thanks for the comments. The main issue that I see is that in SpeechBrain we are supporting and using PyTorch 2.1, which is not yet supported by k2. Do you know when you will release the wheels of PyTorch 2.1?

Reply:
> Do you know when you will release the wheels of PyTorch 2.1?


Please wait for 1 or 2 days.

---

I just built a wheel for torch 2.1.0. The rest versions are still being built by the CI.

https://k2-fsa.github.io/k2/cuda.html

![Screenshot 2023-10-21 at 09 58 21](https://github.com/speechbrain/speechbrain/assets/5284924/f1751364-7739-4dca-bf50-d765e447c4d2)

Reply:
Could you please @csukuangfj ping me when it's ready to be used? I will definitely solve our current CI/tests issues with k2.

Reply:
Now it is ready.

I just wrote a colab notebook to show how to
install a CPU version as well as a CUDA version of k2 with torch 2.1.

https://colab.research.google.com/drive/1mMuNLOwViYI7uyj9WifO02xF2KPBOcHH?usp=sharing

Please leave a comment if you have any issues about the installation.

Reply:
Hello @csukuangfj, 

Thanks for your work! I was able to use pytorch 2.1.0 and k2 on our CI server :)